File: backend/app/routers/persons.py
"""Person-related API endpoints."""
import os
from datetime import datetime
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import FileResponse
from motor.motor_asyncio import AsyncIOMotorDatabase
from bson import ObjectId

from ..database import get_db, to_object_id
from ..models import PersonDocument, person_from_doc, face_from_doc, image_from_doc
from ..schemas import (
    PersonResponse,
    PersonDetail,
    PersonUpdate,
    PersonMergeRequest,
)
from ..services.clustering_service import merge_persons, recalculate_all_clusters

router = APIRouter(prefix="/api/persons", tags=["persons"])


async def _person_to_response(person: PersonDocument, db: AsyncIOMotorDatabase) -> dict:
    """Convert PersonDocument to response dict."""
    # Get faces for this person
    faces_cursor = db.faces.find({"person_id": person.id})
    faces = await faces_cursor.to_list(length=10000)
    
    face_count = len(faces)
    photo_count = len(set(face["image_id"] for face in faces))
    
    # Get thumbnail URL from representative face
    thumbnail_url = None
    if person.representative_face_id:
        face = await db.faces.find_one({"_id": to_object_id(person.representative_face_id)})
        if face and face.get("thumbnail_path"):
            thumbnail_url = f"/api/images/faces/{person.representative_face_id}/thumbnail"
    elif faces:
        # Fallback to first face
        first_face = faces[0]
        if first_face.get("thumbnail_path"):
            thumbnail_url = f"/api/images/faces/{str(first_face['_id'])}/thumbnail"
    
    return {
        "id": person.id,
        "name": person.name,
        "display_name": person.display_name,
        "face_count": face_count,
        "photo_count": photo_count,
        "thumbnail_url": thumbnail_url,
        "created_at": person.created_at,
        "updated_at": person.updated_at,
    }


@router.get("", response_model=List[PersonResponse])
async def list_persons(
    skip: int = 0,
    limit: int = 50,
    labeled: Optional[bool] = None,
    search: Optional[str] = None,
    db: AsyncIOMotorDatabase = Depends(get_db),
):
    """
    List all detected persons.
    """
    query = {}
    
    if labeled is True:
        query["name"] = {"$ne": None}
    elif labeled is False:
        query["name"] = None
    
    if search:
        query["name"] = {"$regex": search, "$options": "i"}
    
    # Sort: named persons first (name exists), then by created_at descending
    # We use aggregation pipeline for more control
    pipeline = [
        {"$match": query},
        {
            "$addFields": {
                "has_name": {"$cond": [{"$ifNull": ["$name", False]}, 1, 0]}
            }
        },
        {"$sort": {"has_name": -1, "created_at": -1}},
        {"$skip": skip},
        {"$limit": limit}
    ]
    
    cursor = db.persons.aggregate(pipeline)
    persons = await cursor.to_list(length=limit)
    
    result = []
    for doc in persons:
        person = person_from_doc(doc)
        result.append(await _person_to_response(person, db))
    
    return result


@router.get("/{person_id}", response_model=PersonDetail)
async def get_person(person_id: str, db: AsyncIOMotorDatabase = Depends(get_db)):
    """Get person details including all their faces."""
    oid = to_object_id(person_id)
    if not oid:
        raise HTTPException(status_code=400, detail="Invalid person ID")
    
    doc = await db.persons.find_one({"_id": oid})
    if not doc:
        raise HTTPException(status_code=404, detail="Person not found")
    
    person = person_from_doc(doc)
    
    # Get faces
    faces_cursor = db.faces.find({"person_id": person_id})
    faces = await faces_cursor.to_list(length=10000)
    
    response = await _person_to_response(person, db)
    response["faces"] = [
        {
            "id": str(face["_id"]),
            "bbox": {
                "top": face["bbox_top"],
                "right": face["bbox_right"],
                "bottom": face["bbox_bottom"],
                "left": face["bbox_left"],
            },
            "thumbnail_url": f"/api/images/faces/{str(face['_id'])}/thumbnail" if face.get("thumbnail_path") else None,
            "person_id": face.get("person_id"),
            "image_id": face["image_id"],
            "created_at": face.get("created_at"),
        }
        for face in faces
    ]
    
    return response


@router.put("/{person_id}", response_model=PersonResponse)
async def update_person(
    person_id: str,
    update: PersonUpdate,
    db: AsyncIOMotorDatabase = Depends(get_db),
):
    """Update a person's name/label."""
    oid = to_object_id(person_id)
    if not oid:
        raise HTTPException(status_code=400, detail="Invalid person ID")
    
    doc = await db.persons.find_one({"_id": oid})
    if not doc:
        raise HTTPException(status_code=404, detail="Person not found")
    
    # Update name
    name = update.name.strip() if update.name and update.name.strip() else None
    await db.persons.update_one(
        {"_id": oid},
        {"$set": {"name": name, "updated_at": datetime.utcnow()}}
    )
    
    # Fetch updated document
    doc = await db.persons.find_one({"_id": oid})
    person = person_from_doc(doc)
    
    return await _person_to_response(person, db)


@router.delete("/{person_id}")
async def delete_person(person_id: str, db: AsyncIOMotorDatabase = Depends(get_db)):
    """
    Delete a person.
    
    Their faces will become unassigned and may be re-clustered.
    """
    oid = to_object_id(person_id)
    if not oid:
        raise HTTPException(status_code=400, detail="Invalid person ID")
    
    doc = await db.persons.find_one({"_id": oid})
    if not doc:
        raise HTTPException(status_code=404, detail="Person not found")
    
    # Unassign all faces
    await db.faces.update_many(
        {"person_id": person_id},
        {"$set": {"person_id": None}}
    )
    
    # Delete person
    await db.persons.delete_one({"_id": oid})
    
    return {"message": "Person deleted successfully"}


@router.get("/{person_id}/photos")
async def get_person_photos(
    person_id: str,
    skip: int = 0,
    limit: int = 50,
    db: AsyncIOMotorDatabase = Depends(get_db),
):
    """Get all photos containing a specific person."""
    oid = to_object_id(person_id)
    if not oid:
        raise HTTPException(status_code=400, detail="Invalid person ID")
    
    doc = await db.persons.find_one({"_id": oid})
    if not doc:
        raise HTTPException(status_code=404, detail="Person not found")
    
    person = person_from_doc(doc)
    
    # Get unique image IDs for this person
    faces_cursor = db.faces.find({"person_id": person_id})
    faces = await faces_cursor.to_list(length=10000)
    
    image_ids = list(set(face["image_id"] for face in faces))
    total_photos = len(image_ids)
    
    # Get images with pagination
    paginated_ids = image_ids[skip:skip + limit]
    images_cursor = db.images.find({"_id": {"$in": [to_object_id(id) for id in paginated_ids]}})
    images = await images_cursor.to_list(length=limit)
    
    # Sort by uploaded_at
    images.sort(key=lambda x: x.get("uploaded_at", datetime.min), reverse=True)
    
    photos = []
    for img in images:
        image = image_from_doc(img)
        
        # Get faces for this person in this image
        person_faces = [f for f in faces if f["image_id"] == image.id]
        
        photos.append({
            "id": image.id,
            "filename": image.filename,
            "original_filename": image.original_filename,
            "thumbnail_url": f"/api/images/{image.id}/thumbnail" if image.thumbnail_path else None,
            "image_url": f"/api/images/{image.id}/file",
            "width": image.width,
            "height": image.height,
            "uploaded_at": image.uploaded_at,
            "faces": [
                {
                    "id": str(face["_id"]),
                    "bbox": {
                        "top": face["bbox_top"],
                        "right": face["bbox_right"],
                        "bottom": face["bbox_bottom"],
                        "left": face["bbox_left"],
                    },
                }
                for face in person_faces
            ],
        })
    
    return {
        "person": await _person_to_response(person, db),
        "total_photos": total_photos,
        "photos": photos,
    }


@router.post("/merge")
async def merge_person_clusters(
    request: PersonMergeRequest,
    db: AsyncIOMotorDatabase = Depends(get_db),
):
    """
    Merge two person clusters.
    """
    if request.source_person_id == request.target_person_id:
        raise HTTPException(
            status_code=400, 
            detail="Source and target person IDs must be different"
        )
    
    success = await merge_persons(db, request.source_person_id, request.target_person_id)
    
    if not success:
        raise HTTPException(
            status_code=404,
            detail="One or both persons not found"
        )
    
    # Get updated target person
    doc = await db.persons.find_one({"_id": to_object_id(request.target_person_id)})
    
    if doc:
        person = person_from_doc(doc)
        return {
            "message": "Persons merged successfully",
            "person": await _person_to_response(person, db),
        }
    
    return {"message": "Persons merged successfully", "person": None}


@router.post("/recluster")
async def recluster_all_faces(db: AsyncIOMotorDatabase = Depends(get_db)):
    """
    Recalculate all face clusters from scratch.
    """
    stats = await recalculate_all_clusters(db)
    
    return {
        "message": "Reclustering complete",
        "stats": stats,
    }


@router.get("/{person_id}/thumbnail")
async def get_person_thumbnail(person_id: str, db: AsyncIOMotorDatabase = Depends(get_db)):
    """Get the representative face thumbnail for a person."""
    oid = to_object_id(person_id)
    if not oid:
        raise HTTPException(status_code=400, detail="Invalid person ID")
    
    doc = await db.persons.find_one({"_id": oid})
    if not doc:
        raise HTTPException(status_code=404, detail="Person not found")
    
    person = person_from_doc(doc)
    
    # Try representative face first
    face = None
    if person.representative_face_id:
        face_doc = await db.faces.find_one({"_id": to_object_id(person.representative_face_id)})
        if face_doc:
            face = face_doc
    
    # Fallback to first face
    if not face:
        face = await db.faces.find_one({"person_id": person_id})
    
    if not face or not face.get("thumbnail_path"):
        raise HTTPException(status_code=404, detail="No thumbnail available")
    
    if not os.path.exists(face["thumbnail_path"]):
        raise HTTPException(status_code=404, detail="Thumbnail file not found")
    
    return FileResponse(face["thumbnail_path"], media_type="image/jpeg")

File: backend/src/routes/persons.ts
import { Hono } from 'hono'
import { ObjectId } from 'mongodb'
import { getDb } from '../db'
import { adminAuth } from '../middleware/auth'

const persons = new Hono()

// Helper to convert PersonDocument to response
const personToResponse = async (person: any, db: any) => {
  const faces = await db.collection('faces').find({ person_id: person._id.toString() }).toArray()
  
  const faceCount = faces.length
  const photoCount = new Set(faces.map((f: any) => f.image_id)).size
  
  let thumbnailUrl = null
  if (person.representative_face_id) {
    const face = await db.collection('faces').findOne({ _id: new ObjectId(person.representative_face_id) })
    if (face && face.thumbnail_path) {
      thumbnailUrl = `/api/images/faces/${person.representative_face_id}/thumbnail`
    }
  } else if (faces.length > 0) {
    const firstFace = faces[0]
    if (firstFace.thumbnail_path) {
      thumbnailUrl = `/api/images/faces/${firstFace._id.toString()}/thumbnail`
    }
  }
  
  return {
    id: person._id.toString(),
    name: person.name,
    display_name: person.display_name, // Assuming this exists in DB or model
    face_count: faceCount,
    photo_count: photoCount,
    thumbnail_url: thumbnailUrl,
    created_at: person.created_at,
    updated_at: person.updated_at,
  }
}

// List persons
persons.get('/', async (c) => {
  const db = await getDb()
  const skip = parseInt(c.req.query('skip') || '0')
  const limit = parseInt(c.req.query('limit') || '50')
  const labeled = c.req.query('labeled')
  const search = c.req.query('search')
  
  const query: any = {}
  
  if (labeled === 'true') {
    query.name = { $ne: null }
  } else if (labeled === 'false') {
    query.name = null
  }
  
  if (search) {
    query.name = { $regex: search, $options: 'i' }
  }
  
  const pipeline = [
    { $match: query },
    {
      $addFields: {
        has_name: { $cond: [{ $ifNull: ["$name", false] }, 1, 0] }
      }
    },
    { $sort: { has_name: -1, created_at: -1 } },
    { $skip: skip },
    { $limit: limit }
  ]
  
  const docs = await db.collection('persons').aggregate(pipeline).toArray()
  
  const result = []
  for (const doc of docs) {
    result.push(await personToResponse(doc, db))
  }
  
  return c.json(result)
})

// Get person details
persons.get('/:id', async (c) => {
  const db = await getDb()
  const id = c.req.param('id')
  
  if (!ObjectId.isValid(id)) {
    return c.json({ error: 'Invalid person ID' }, 400)
  }
  
  const doc = await db.collection('persons').findOne({ _id: new ObjectId(id) })
  if (!doc) {
    return c.json({ error: 'Person not found' }, 404)
  }
  
  const response: any = await personToResponse(doc, db)
  
  const faces = await db.collection('faces').find({ person_id: id }).toArray()
  
  response.faces = faces.map((face: any) => ({
    id: face._id.toString(),
    bbox: {
      top: face.bbox_top,
      right: face.bbox_right,
      bottom: face.bbox_bottom,
      left: face.bbox_left,
    },
    thumbnail_url: face.thumbnail_path ? `/api/images/faces/${face._id.toString()}/thumbnail` : null,
    person_id: face.person_id,
    image_id: face.image_id,
    created_at: face.created_at,
  }))
  
  return c.json(response)
})

// Update person
persons.put('/:id', adminAuth, async (c) => {
  const db = await getDb()
  const id = c.req.param('id')
  const body = await c.req.json()
  
  if (!ObjectId.isValid(id)) {
    return c.json({ error: 'Invalid person ID' }, 400)
  }
  
  const doc = await db.collection('persons').findOne({ _id: new ObjectId(id) })
  if (!doc) {
    return c.json({ error: 'Person not found' }, 404)
  }
  
  const name = body.name && body.name.trim() ? body.name.trim() : null
  
  await db.collection('persons').updateOne(
    { _id: new ObjectId(id) },
    { $set: { name, updated_at: new Date() } }
  )
  
  const updatedDoc = await db.collection('persons').findOne({ _id: new ObjectId(id) })
  return c.json(await personToResponse(updatedDoc, db))
})

// Delete person
persons.delete('/:id', adminAuth, async (c) => {
  const db = await getDb()
  const id = c.req.param('id')
  
  if (!ObjectId.isValid(id)) {
    return c.json({ error: 'Invalid person ID' }, 400)
  }
  
  const doc = await db.collection('persons').findOne({ _id: new ObjectId(id) })
  if (!doc) {
    return c.json({ error: 'Person not found' }, 404)
  }
  
  // Unassign faces
  await db.collection('faces').updateMany(
    { person_id: id },
    { $set: { person_id: null } }
  )
  
  await db.collection('persons').deleteOne({ _id: new ObjectId(id) })
  
  return c.json({ message: 'Person deleted successfully' })
})

// Get person photos
persons.get('/:id/photos', async (c) => {
  const db = await getDb()
  const id = c.req.param('id')
  const skip = parseInt(c.req.query('skip') || '0')
  const limit = parseInt(c.req.query('limit') || '50')
  
  if (!ObjectId.isValid(id)) {
    return c.json({ error: 'Invalid person ID' }, 400)
  }
  
  const doc = await db.collection('persons').findOne({ _id: new ObjectId(id) })
  if (!doc) {
    return c.json({ error: 'Person not found' }, 404)
  }
  
  const faces = await db.collection('faces').find({ person_id: id }).toArray()
  const imageIds = [...new Set(faces.map((f: any) => f.image_id))]
  const totalPhotos = imageIds.length
  
  const paginatedIds = imageIds.slice(skip, skip + limit)
  const objectIds = paginatedIds.map((id: string) => new ObjectId(id))
  
  const images = await db.collection('images').find({ _id: { $in: objectIds } }).toArray()
  
  // Sort by uploaded_at desc
  images.sort((a: any, b: any) => {
    const dateA = a.uploaded_at ? new Date(a.uploaded_at).getTime() : 0
    const dateB = b.uploaded_at ? new Date(b.uploaded_at).getTime() : 0
    return dateB - dateA
  })
  
  const photos = images.map((image: any) => {
    const personFaces = faces.filter((f: any) => f.image_id === image._id.toString())
    
    return {
      id: image._id.toString(),
      filename: image.filename,
      original_filename: image.original_filename,
      thumbnail_url: image.thumbnail_path ? `/api/images/${image._id.toString()}/thumbnail` : null,
      image_url: `/api/images/${image._id.toString()}/file`,
      width: image.width,
      height: image.height,
      uploaded_at: image.uploaded_at,
      faces: personFaces.map((face: any) => ({
        id: face._id.toString(),
        bbox: {
          top: face.bbox_top,
          right: face.bbox_right,
          bottom: face.bbox_bottom,
          left: face.bbox_left,
        }
      }))
    }
  })
  
  return c.json({
    person: await personToResponse(doc, db),
    total_photos: totalPhotos,
    photos
  })
})

// Merge persons
persons.post('/merge', adminAuth, async (c) => {
  const db = await getDb()
  const body = await c.req.json()
  const { source_person_id, target_person_id } = body
  
  if (source_person_id === target_person_id) {
    return c.json({ error: 'Source and target person IDs must be different' }, 400)
  }
  
  const sourcePerson = await db.collection('persons').findOne({ _id: new ObjectId(source_person_id) })
  const targetPerson = await db.collection('persons').findOne({ _id: new ObjectId(target_person_id) })
  
  if (!sourcePerson || !targetPerson) {
    return c.json({ error: 'One or both persons not found' }, 404)
  }
  
  // Move faces from source to target
  await db.collection('faces').updateMany(
    { person_id: source_person_id },
    { $set: { person_id: target_person_id } }
  )
  
  // Delete source person
  await db.collection('persons').deleteOne({ _id: new ObjectId(source_person_id) })
  
  // Update target person timestamp
  await db.collection('persons').updateOne(
    { _id: new ObjectId(target_person_id) },
    { $set: { updated_at: new Date() } }
  )
  
  const updatedTarget = await db.collection('persons').findOne({ _id: new ObjectId(target_person_id) })
  
  return c.json({
    message: 'Persons merged successfully',
    person: await personToResponse(updatedTarget, db)
  })
})

export default persons
